src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -188,10 +189,49 @@ public Stream<MemorySegment> elements(MemoryLayout elementLayout) {
         return StreamSupport.stream(spliterator(elementLayout), false);
     }
 
+    // FILL_NATIVE_THRESHOLD must be a power of two and should be greater than 2^3
+    // Update the value for Aarch64 once 8338975 is fixed.
+    private static final long FILL_NATIVE_THRESHOLD = 1L << (Architecture.isAARCH64() ? 10 : 5);
+
     @Override
-    public final MemorySegment fill(byte value){
-        checkAccess(0, length, false);
-        SCOPED_MEMORY_ACCESS.setMemory(sessionImpl(), unsafeGetBase(), unsafeGetOffset(), length, value);
+    public final MemorySegment fill(byte value) {
+        checkReadOnly(false);
+        // 0...0X...XXXX implies: 0 <= length < FILL_NATIVE_LIMIT
+        if ((length & -FILL_NATIVE_THRESHOLD) == 0) {
```

Comment:
Looks great - few notes:
* adding `@ForceInline` here might help
* I'd suggest to move the `length == 0` outside in its own `if` branch (and drop the `return` from there). E.g.

```
if (length == 0) { ... }
else if (length & -FILL...) { ... }
else { ... }
return this;
```